### PR TITLE
New collector for garden containers

### PR DESCRIPTION
**Pull request by zippolyte**
### What does this PR do?

Adds a collector for garden containers, responsible for running apps on cloudfoundry
It collects the list of running containers by querying a local Garden API, then fills the metrics by parsing the cgroups stats.
To collect additional metadata such as the application name and everything, it connects to the cluster agent that maintains a cache of the metadata coming from a central BBS API. This is done to avoid hammering that BBS API. See #4935 for more details

### CODE REVIEW

**Comment by vboulineau** [comment]
You might want to integrate with on going work to revamp the metrics providers: https://github.com/DataDog/datadog-agent/pull/4885.

I don't know how PCF works, but you cannot use the existing cgroup implementation to retrieve stats?

**Comment by zippolyte** [comment]
Thanks for your comment @vboulineau 
I went ahead and switched to using the already existing cgroup implementation to collect metrics, and it seems to work better. I'll try to include your work next.

File: cmd/cluster-agent/api/server.go
Comment (ahmed-mez) [inline cmd/cluster-agent/api/server.go]: shouldn't node agents query this by declaring their node names and only get tags that correspond to the apps running on that specific node? we can avoid sending huge payloads and simplify the tag-filtering logic in the node agents

File: cmd/cluster-agent/api/v1/install.go
Comment (ahmed-mez) [inline cmd/cluster-agent/api/v1/install.go]: `getCFAppMetadata` doesn't match the function name

File: pkg/tagger/collectors/garden_extract.go
Comment (ahmed-mez) [inline pkg/tagger/collectors/garden_extract.go]: not familiar with cloud foundry, but is it intended to have these two tags with the same value `gardenContainer.Handle()` ?

File: cmd/cluster-agent/api/server.go
Comment (zippolyte) [reply cmd/cluster-agent/api/server.go]: Good point, I'll add that in

File: pkg/tagger/collectors/garden_extract.go
Comment (zippolyte) [reply pkg/tagger/collectors/garden_extract.go]: This `container_name` tag value is a fallback for when the DCA agent isn't enabled, just so that we don't have an empty container name (empty container name can cause issues in the UI). 
Let me know if that's not a proper option

File: pkg/tagger/collectors/garden_main.go
Comment (ahmed-mez) [inline pkg/tagger/collectors/garden_main.go]: are we sure `containers.SplitEntityName(entity)` returns the container id string that corresponds to ids in the tags payload?

File: pkg/util/cloudfoundry/bbscache.go
Comment (ahmed-mez) [inline pkg/util/cloudfoundry/bbscache.go]: would be super useful we can map tags by nodes (VMs), each node agent puts its node name in the query path and gets only a subset of tags of apps running on that same node

File: cmd/cluster-agent/api/server.go
Comment (ahmed-mez) [reply cmd/cluster-agent/api/server.go]: thanks, please make sure you validate the url by checking the prefix and the value of `len(strings.Split(path, "/"))`

File: pkg/tagger/collectors/garden_extract.go
Comment (ahmed-mez) [reply pkg/tagger/collectors/garden_extract.go]: sounds good, thanks for the clarification - is there a way to get the container name from the local api? wouldn't expect to see a GUID as container name in the UI

File: pkg/tagger/collectors/garden_main.go
Comment (zippolyte) [reply pkg/tagger/collectors/garden_main.go]: Yeah I create the entity name with `containers.BuildTaggerEntityName`, passing in the handle of the container, which is the container Instance GUID that we get in the BBS Cache object

File: pkg/tagger/collectors/garden_extract.go
Comment (zippolyte) [reply pkg/tagger/collectors/garden_extract.go]: AFAICS, there is no way to have the metadata from the local API, we have to query the BBS API to get it. Since it's not something we want to do for each node agent for performance reasons, this is the fallback that I saw.
However, we'll make it so that using our tile to deploy the agent in WMWare CloudFoundry will automatically deploy an instance of the DCA whenever process collection for containers is enabled, and we'll document that it's needed to have the best experience when manually deploying agents using open source CloudFoundry.
This is mostly a safeguard in case the cluster agent isn't enable. I thought it would be better than a hard fail and no container collection at all

File: pkg/util/cloudfoundry/garden_linux.go
Comment (ahmed-mez) [inline pkg/util/cloudfoundry/garden_linux.go]: I think the `Name` field is used to show the container in the live container view, not sure how this is going to be handled if we send it empty.
cc @DataDog/processes might know more about this

File: pkg/tagger/collectors/garden_main.go
Comment (ahmed-mez) [reply pkg/tagger/collectors/garden_main.go]: 💯

File: pkg/util/containers/metrics/cgroup_detect.go
Comment (ahmed-mez) [inline pkg/util/containers/metrics/cgroup_detect.go]: could you please add a comment to explain why we extended this? could you also add some test cases for garden containers in `cgroup_detect_test.go` ?

File: pkg/tagger/collectors/garden_main.go
Comment (bkabrda) [inline pkg/tagger/collectors/garden_main.go]: Do you need this condition when you `return NoCollection, err` anyway?

**Review by bkabrda — COMMENTED** [review]
This looks really good. I just left one minor comment inline.

File: pkg/util/cloudfoundry/garden_linux.go
Comment (zippolyte) [reply pkg/util/cloudfoundry/garden_linux.go]: They told me it's not used anymore, and that they get the name from the `container_name` tag

File: pkg/tagger/collectors/garden_main.go
Comment (zippolyte) [reply pkg/tagger/collectors/garden_main.go]: nope I don't

File: pkg/util/cloudfoundry/garden_linux.go
Comment (ahmed-mez) [reply pkg/util/cloudfoundry/garden_linux.go]: didn't know that,  thanks 👍

**Review by shang-wang — DISMISSED** [review]
The container collector part looks good to me!

**Comment by zippolyte** [comment]
@ahmed-mez, @bkabrda, @shang-wang I've updated the PR after #4935 was merged, mind giving it a second look ? Thanks

File: cmd/cluster-agent/api/v1/install.go
Comment (ahmed-mez) [inline cmd/cluster-agent/api/v1/install.go]: it's a bit confusing to see cf and k8s handlers in the same file, can we at least move them to separate files? e.g `kubernetes_metadata.go` and `cloudfoundry_metatada.go` with an `install` in every file and we call them in  `Install` the same way we call `installClusterCheckEndpoints` and `installEndpointsCheckEndpoints` wdyt?
we can also limit the installed endpoints depending on the environment, there is no need to expose k8s tag endpoints when cf enabled and vice versa - this could reduce the security risks as these endpoints are publicly accessible

File: pkg/tagger/collectors/garden_extract.go
Comment (ahmed-mez) [inline pkg/tagger/collectors/garden_extract.go]:
```suggestion
		log.Debug("Cluster agent not enabled or misconfigured, tagging CF app with container id only")
```

File: pkg/tagger/collectors/garden_main.go
Comment (ahmed-mez) [inline pkg/tagger/collectors/garden_main.go]:
```suggestion
				log.Error("Permanent failure in communication with the cluster agent")
```

File: pkg/util/cloudfoundry/bbscache.go
Comment (ahmed-mez) [inline pkg/util/cloudfoundry/bbscache.go]: should this method return and error if `instanceGUID` is not found?

File: pkg/util/cloudfoundry/bbscache.go
Comment (ahmed-mez) [inline pkg/util/cloudfoundry/bbscache.go]: should this method return and error if `appGUID ` is not found?

File: cmd/cluster-agent/api/v1/install.go
Comment (zippolyte) [reply cmd/cluster-agent/api/v1/install.go]: How do I limit the installed endpoints ?

File: pkg/util/cloudfoundry/bbscache.go
Comment (ahmed-mez) [inline pkg/util/cloudfoundry/bbscache.go]: does the bbs api offer the possibility to query LRPs by node name directly? might be super useful to avoid extra processing here

File: cmd/cluster-agent/api/v1/install.go
Comment (ahmed-mez) [reply cmd/cluster-agent/api/v1/install.go]: we can check the config of the agent to know which mode is activated, although this makes wonder why we're keeping this pkg under `cluster-agent` - maybe we need to move it in a common package and both cluster agent can plug to it

File: pkg/util/cloudfoundry/garden.go
Comment (ahmed-mez) [inline pkg/util/cloudfoundry/garden.go]:
```suggestion
		log.Debugf("Could not initiate connection to garden server %s using network %s: %s", address, network, err)
```

File: pkg/util/cloudfoundry/garden_linux.go
Comment (ahmed-mez) [inline pkg/util/cloudfoundry/garden_linux.go]: we can log something here

File: pkg/util/cloudfoundry/garden_linux.go
Comment (ahmed-mez) [inline pkg/util/cloudfoundry/garden_linux.go]: we can log something here

File: pkg/util/cloudfoundry/garden_linux.go
Comment (ahmed-mez) [inline pkg/util/cloudfoundry/garden_linux.go]: these errors will result zero values in the live container view, I'd rather log these as warnings or errors

File: pkg/util/clusteragent/clusteragent.go
Comment (ahmed-mez) [inline pkg/util/clusteragent/clusteragent.go]: can we unit test this method in `clusteragent_test.go` please?

File: pkg/util/cloudfoundry/bbscache.go
Comment (zippolyte) [reply pkg/util/cloudfoundry/bbscache.go]: We're not reading from the BBS API here, but from our local cache. That said, I can probably separate by cell ID when I read all the actual LRPs and store them in our cache for easier lookup

Comment (ahmed-mez) [reply pkg/util/cloudfoundry/bbscache.go]: sounds good 👍

File: pkg/util/cloudfoundry/garden_linux.go
Comment (zippolyte) [reply pkg/util/cloudfoundry/garden_linux.go]: I reused the same log level as for `docker` collector though, so I'd keep it the same for consistency. WDYT ? https://github.com/DataDog/datadog-agent/blob/32a13a35e57a570943ba555e82a096c97a426571/pkg/util/docker/containers.go#L136-L140

Comment (ahmed-mez) [reply pkg/util/cloudfoundry/garden_linux.go]: fair enough 👍  the error returned by `metrics.ScrapeAllCgroups()` before the loop should be enough to report cgroup related problems

**Review by ahmed-mez — DISMISSED** [review]

File: pkg/util/cloudfoundry/bbscache.go
Comment (bkabrda) [inline pkg/util/cloudfoundry/bbscache.go]: Should this error say "LRPs for cell"?

File: pkg/util/cloudfoundry/bbscache.go
Comment (bkabrda) [inline pkg/util/cloudfoundry/bbscache.go]: It looks like this block is the same here and then in the `else` clause below. Maybe you should extract it and put it just once after this condition?

File: pkg/util/cloudfoundry/bbscache.go
Comment (bkabrda) [inline pkg/util/cloudfoundry/bbscache.go]: Do you think it would make more sense to have this pre-computed for all cells in the same way the actual/desired LRPs are and then just return result? Doing all this dynamically with potentially thousands of nodes polling in short interval could be very CPU-intensive when recomputing all the time.

File: pkg/util/cloudfoundry/types.go
Comment (bkabrda) [inline pkg/util/cloudfoundry/types.go]: What about unmarshalling and storing exactly the fields that we need later? IMO it would be sufficient then to have the return value `map[string]string` with just the relevant information (on clusters with thousands of apps that might save some memory).

File: pkg/tagger/collectors/garden_extract.go
Comment (bkabrda) [inline pkg/tagger/collectors/garden_extract.go]: We should probably introduce constants for `container_name` and `app_instance_guid` to prevent typos or inconsistencies in future.

**Review by bkabrda — CHANGES_REQUESTED** [review]
This looks really good, I left couple comments inline, mostly about caching some data that is now be computed dynamically, but also couple other minor things to improve.

File: pkg/util/cloudfoundry/bbscache.go
Comment (zippolyte) [reply pkg/util/cloudfoundry/bbscache.go]: doh, indeed, thanks !

**Review by bkabrda — DISMISSED** [review]

**Review by bkabrda — DISMISSED** [review]

**Review by ahmed-mez — DISMISSED** [review]

**Review by shang-wang — DISMISSED** [review]

File: pkg/config/config.go
Comment (olivielpeau) [inline pkg/config/config.go]: any reason not to document these options in `config_template.yaml`?

File: pkg/util/containers/collectors/cloudoundry.go
Comment (olivielpeau) [inline pkg/util/containers/collectors/cloudoundry.go]: this file has a typo in its name

**Review by olivielpeau — DISMISSED** [review]
Only reviewed very briefly, didn't really review the actual logic. A couple of inline nits, can be addressed in separate PR if that's easier for you.

Can you update the CODEOWNERS file? Typically the cloudfoundry-related utils should be owned by intg-tools-libs. Again, can be done in separate PR.

File: pkg/config/config.go
Comment (zippolyte) [reply pkg/config/config.go]: nope, oversight

File: pkg/util/containers/collectors/cloudoundry.go
Comment (zippolyte) [reply pkg/util/containers/collectors/cloudoundry.go]: 🙈

File: pkg/config/config.go
Comment (zippolyte) [reply pkg/config/config.go]: making a note for once this is merged

**Review by bkabrda — DISMISSED** [review]

**Review by ahmed-mez — DISMISSED** [review]

File: pkg/util/cloudfoundry/garden_linux.go
Comment (vboulineau) [inline pkg/util/cloudfoundry/garden_linux.go]: This call is not necessary in itself as `Prefetch` was already done.
To avoid doing 2 `Prefetch` and 2 loops over containers, it might be best to mimic the changes that were done in `DockerUtil`.

File: pkg/util/cloudfoundry/garden_linux.go
Comment (vboulineau) [inline pkg/util/cloudfoundry/garden_linux.go]: Except if `"code.cloudfoundry.org/garden"` has some `linux` build flags, you should not need to have a `_linux` and `_windows` files as `ContainerProvider` does not depend on a specific OS anymore.

File: pkg/util/cloudfoundry/garden_linux.go
Comment (vboulineau) [inline pkg/util/cloudfoundry/garden_linux.go]: The idea is to put this import where we know the OS we'll be running on, usually around the `main` file, if possible

**Comment by zippolyte** [comment]
@olivielpeau @vboulineau @bkabrda @ahmed-mez @DataDog/agent-platform @DataDog/networks 
Could you please have a final review on this ? I think I've addressed everything, would love to get this merged pretty soon so I can finish this work.

**Review by vboulineau — DISMISSED** [review]

File: pkg/util/cloudfoundry/garden.go
Comment (ahmed-mez) [inline pkg/util/cloudfoundry/garden.go:131]:
```suggestion
		if container.State != containers.ContainerActiveState {
			log.Debugf("Container %s not in state %s, skipping", container.ID[:12], containers.ContainerActiveState)
			continue
		}
		if !providers.ContainerImpl().ContainerExists(container.ID) {
			log.Debugf("Container %s not found, skipping", container.ID[:12])
			continue
		}
```

File: pkg/util/cloudfoundry/garden.go
Comment (ahmed-mez) [inline pkg/util/cloudfoundry/garden.go:149]: same, let's separate these to make sure we log why we're skipping correctly

**Review by albertvaka — DISMISSED** [review]
👍 for changes to gopkg (although it will conflict with the move to go modules)

**Review by ahmed-mez — DISMISSED** [review]
2 small comments, great work!

**Review by ahmed-mez — APPROVED** [review]

**Review by olivielpeau — APPROVED** [review]

**Review by albertvaka — APPROVED** [review]
Re-approving

**Review by bkabrda — APPROVED** [review]

**Comment by sarina-dd** [comment]
👍 docs

**Review by sarina-dd — APPROVED** [review]

**Review by stb2132 — APPROVED** [review]

**Review by leeavital — APPROVED** [review]

**Pull request merged**
